// @frontegg/rest-api/node/applications/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assignUserToApplications = assignUserToApplications;
exports.getApplicationsData = getApplicationsData;
exports.getTenantsApplications = getTenantsApplications;
exports.getUserApplicationsId = getUserApplicationsId;
exports.getUsersApplicationsId = getUsersApplicationsId;
exports.getUsersOfApplications = getUsersOfApplications;
exports.unassignUserFromApplications = unassignUserFromApplications;

var _constants = require("../constants");

var _fetch = require("../fetch");

async function getUserApplicationsId({
  userId
}) {
  return (0, _fetch.Get)(`${_constants.urls.identity.applications.v1}/${userId}/apps`);
}

async function getUsersApplicationsId({
  userIds
}) {
  return (0, _fetch.Get)(`${_constants.urls.identity.applications.v1}/users-apps`, {
    userIds: userIds.join(',')
  });
}

async function getTenantsApplications() {
  return (0, _fetch.Get)(_constants.urls.applications.tenant.v1);
}

async function getApplicationsData({
  appIds,
  includeFreeAccess = true
}) {
  return (0, _fetch.Get)(_constants.urls.applications.v1, {
    ids: appIds.join(','),
    _includeFreeAccess: includeFreeAccess
  });
}

async function assignUserToApplications(body) {
  return (0, _fetch.Post)(`${_constants.urls.identity.applications.v1}/apps-user`, body);
}

async function unassignUserFromApplications(body) {
  return (0, _fetch.Delete)(`${_constants.urls.identity.applications.v1}/user-apps`, body);
}

async function getUsersOfApplications({
  appIds
}) {
  return (0, _fetch.Get)(`${_constants.urls.identity.applications.v1}/apps-users`, {
    appIds: appIds.join(',')
  });
}
