// @frontegg/redux-store/node/auth/ApplicationsState/saga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.applicationsSaga = applicationsSaga;
exports.assignNewUserToApps = assignNewUserToApps;
exports.assignUserToApps = assignUserToApps;
exports.loadAccountApplications = loadAccountApplications;
exports.loadMultipleUsersApplications = loadMultipleUsersApplications;
exports.loadUserApplications = loadUserApplications;
exports.mapUsersWithApplicationData = mapUsersWithApplicationData;
exports.unassignUserFromApps = unassignUserFromApps;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _restApi = require("@frontegg/rest-api");
var _effects = require("redux-saga/effects");
var _utils = require("../../utils");
var _reducer = require("../reducer");
var _helpers = require("../../helpers");
var _getFlattenAppIds = require("./utils/getFlattenAppIds");
const ASSIGNMENT_DEFAULT_ERROR_MESSAGE = 'Assignment action failed. Try again or contact support.';
function* loadUserApplications() {
  yield (0, _effects.put)(_reducer.actions.setApplicationsState({
    loading: true,
    fetching: true
  }));
  try {
    const userId = yield (0, _effects.select)(({
      auth
    }) => {
      var _auth$user;
      return auth == null ? void 0 : (_auth$user = auth.user) == null ? void 0 : _auth$user.id;
    });
    const appIds = yield (0, _effects.call)(_restApi.api.applications.getUserApplicationsId, {
      userId
    });
    const userApplications = yield (0, _effects.call)(_restApi.api.applications.getApplicationsData, {
      appIds
    });
    yield (0, _effects.put)(_reducer.actions.setApplicationsState({
      userApplications,
      fetching: false,
      loading: false
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setApplicationsState({
      loading: false,
      fetching: false,
      error: (0, _utils.errorHandler)(e)
    }));
  }
}
function* loadAccountApplications() {
  yield (0, _effects.put)(_reducer.actions.setApplicationsState({
    loading: true,
    fetching: true
  }));
  try {
    const tenantAppIds = yield (0, _effects.call)(_restApi.api.applications.getTenantsApplications);
    const appIds = (0, _getFlattenAppIds.getFlattenAppIds)(tenantAppIds);
    const accountApplications = yield (0, _effects.call)(_restApi.api.applications.getApplicationsData, {
      appIds
    });
    const usersOfApplications = yield (0, _effects.call)(_restApi.api.applications.getUsersOfApplications, {
      appIds
    });
    const accountApplicationsWithUsers = accountApplications.map(app => {
      var _usersOfApplications$;
      return (0, _extends2.default)({}, app, {
        userIds: (_usersOfApplications$ = usersOfApplications.find(u => u.appId === app.id)) == null ? void 0 : _usersOfApplications$.userIds
      });
    });
    yield (0, _effects.put)(_reducer.actions.setApplicationsState({
      accountApplications: accountApplicationsWithUsers,
      fetching: false,
      loading: false
    }));
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setApplicationsState({
      loading: false,
      fetching: false,
      error: (0, _utils.errorHandler)(e)
    }));
  }
}
function* loadMultipleUsersApplications({
  userIds
}) {
  try {
    const usersAppIds = yield (0, _effects.call)(_restApi.api.applications.getUsersApplicationsId, {
      userIds
    });
    const appIds = (0, _getFlattenAppIds.getFlattenAppIds)(usersAppIds);
    const accountApplications = yield (0, _effects.call)(_restApi.api.applications.getApplicationsData, {
      appIds
    });
    return userIds.reduce((acc, userId) => {
      var _usersAppIds$find, _userAppIds$map$filte, _userAppIds$map;
      const userAppIds = (_usersAppIds$find = usersAppIds.find(app => app.userId == userId)) == null ? void 0 : _usersAppIds$find.appIds;
      acc[userId] = [...accountApplications.filter(app => app.accessType === _restApi.ApplicationAccessType.FREE_ACCESS), ...((_userAppIds$map$filte = userAppIds == null ? void 0 : (_userAppIds$map = userAppIds.map(appId => accountApplications.find(app => appId === app.id))) == null ? void 0 : _userAppIds$map.filter(app => !!app)) != null ? _userAppIds$map$filte : [])];
      return acc;
    }, {});
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setApplicationsState({
      error: (0, _utils.errorHandler)(e)
    }));
  }
}
function* assignUserToApps({
  payload
}) {
  const {
    userId,
    appIds,
    callback
  } = payload;
  try {
    var _select, _accountApps$filter;
    const tenantId = yield (0, _effects.select)(({
      auth
    }) => {
      var _auth$user2;
      return auth == null ? void 0 : (_auth$user2 = auth.user) == null ? void 0 : _auth$user2.tenantId;
    });
    yield (0, _effects.call)(_restApi.api.applications.assignUserToApplications, {
      userId,
      appIds,
      tenantId
    });
    const accountApps = yield (_select = (0, _effects.select)(({
      auth
    }) => {
      var _auth$applicationsSta;
      return auth == null ? void 0 : (_auth$applicationsSta = auth.applicationsState) == null ? void 0 : _auth$applicationsSta.accountApplications;
    })) != null ? _select : [];
    const users = yield (0, _effects.select)(({
      auth
    }) => {
      var _auth$teamState;
      return auth == null ? void 0 : (_auth$teamState = auth.teamState) == null ? void 0 : _auth$teamState.users;
    });
    const newApps = (_accountApps$filter = accountApps == null ? void 0 : accountApps.filter(app => appIds.includes(app.id))) != null ? _accountApps$filter : [];
    if (newApps != null && newApps.length) {
      yield (0, _effects.put)(_reducer.actions.setTeamState({
        users: users.map(user => {
          var _user$applications;
          return user.id === userId ? (0, _extends2.default)({}, user, {
            applications: [...((_user$applications = user.applications) != null ? _user$applications : []), ...newApps]
          }) : user;
        })
      }));
    }
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setApplicationsState({
      error: (0, _utils.errorHandler)(e, ASSIGNMENT_DEFAULT_ERROR_MESSAGE)
    }));
    callback == null ? void 0 : callback(false);
  }
}
function* unassignUserFromApps({
  payload
}) {
  const {
    userId,
    appIds,
    callback
  } = payload;
  try {
    var _users$find;
    const tenantId = yield (0, _effects.select)(({
      auth
    }) => {
      var _auth$user3;
      return auth == null ? void 0 : (_auth$user3 = auth.user) == null ? void 0 : _auth$user3.tenantId;
    });
    yield (0, _effects.call)(_restApi.api.applications.unassignUserFromApplications, {
      userId,
      appIds,
      tenantId
    });
    const users = yield (0, _effects.select)(({
      auth
    }) => {
      var _auth$teamState2;
      return auth == null ? void 0 : (_auth$teamState2 = auth.teamState) == null ? void 0 : _auth$teamState2.users;
    });
    const userApps = (_users$find = users.find(user => user.id === userId)) == null ? void 0 : _users$find.applications;
    if (userApps != null && userApps.length && appIds.length) {
      yield (0, _effects.put)(_reducer.actions.setTeamState({
        users: users.map(user => user.id === userId ? (0, _extends2.default)({}, user, {
          applications: userApps.filter(app => !appIds.includes(app.id))
        }) : user)
      }));
    }
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield (0, _effects.put)(_reducer.actions.setApplicationsState({
      error: (0, _utils.errorHandler)(e, ASSIGNMENT_DEFAULT_ERROR_MESSAGE)
    }));
    callback == null ? void 0 : callback(false);
  }
}
function* assignNewUserToApps({
  appIds,
  user
}) {
  var _apps$filter;
  const [multiAppsFlag] = yield (0, _effects.call)(_helpers.getFeatureFlags, ['multi-apps-admin-portal']);
  if (!multiAppsFlag) {
    return;
  }
  const apps = yield (0, _effects.select)(({
    auth
  }) => {
    var _auth$applicationsSta2;
    return auth == null ? void 0 : (_auth$applicationsSta2 = auth.applicationsState) == null ? void 0 : _auth$applicationsSta2.accountApplications;
  });
  const defaultApps = (_apps$filter = apps == null ? void 0 : apps.filter(app => app.accessType === _restApi.ApplicationAccessType.FREE_ACCESS)) != null ? _apps$filter : [];
  if (appIds != null && appIds.length) {
    const tenantId = yield (0, _effects.select)(({
      auth
    }) => {
      var _auth$user4;
      return auth == null ? void 0 : (_auth$user4 = auth.user) == null ? void 0 : _auth$user4.tenantId;
    });
    yield (0, _effects.call)(_restApi.api.applications.assignUserToApplications, {
      userId: user.id,
      appIds,
      tenantId
    });
    return appIds.map(appId => {
      var _apps$find;
      return (_apps$find = apps == null ? void 0 : apps.find(app => app.id === appId)) != null ? _apps$find : [];
    }).concat(defaultApps);
  } else {
    return defaultApps;
  }
}
function* mapUsersWithApplicationData({
  shouldLoadApps,
  users
}) {
  const [multiAppsFlag] = yield (0, _effects.call)(_helpers.getFeatureFlags, ['multi-apps-admin-portal']);
  if (!multiAppsFlag || !shouldLoadApps) {
    return users;
  }
  const userIds = users.map(user => user.id);
  const userApps = yield (0, _effects.call)(loadMultipleUsersApplications, {
    userIds
  });
  if (!userApps) {
    return users;
  }
  return users.map(user => userApps != null && userApps[user.id] ? (0, _extends2.default)({}, user, {
    applications: userApps[user.id]
  }) : user);
}
function* applicationsSaga() {
  yield (0, _effects.takeLeading)(_reducer.actions.loadUserApplications, loadUserApplications);
  yield (0, _effects.takeLeading)(_reducer.actions.loadAccountApplications, loadAccountApplications);
  yield (0, _effects.takeLeading)(_reducer.actions.assignUserToApps, assignUserToApps);
  yield (0, _effects.takeLeading)(_reducer.actions.unassignUserFromApps, unassignUserFromApps);
}
