// @frontegg/redux-store/auth/ImpersonationState/saga.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["callback", "events"];
import { api } from '@frontegg/rest-api';
import { call, delay, put, takeLeading } from 'redux-saga/effects';
import { afterAuthNavigation } from '../LoginState/saga';
import { actions } from '../reducer';
import { ImpersonateStep } from './interfaces';
import { errorHandler } from '../../utils';
function* impersonate(_ref) {
  let {
      payload: {
        callback
      }
    } = _ref,
    payload = _objectWithoutPropertiesLoose(_ref.payload, _excluded);
  yield put(actions.setImpersonateState({
    loading: true
  }));
  try {
    yield call(api.impersonate.impersonate, payload);
    yield put(actions.setImpersonateState({
      step: ImpersonateStep.success
    }));
    yield delay(1000);
    /**
     * We want to avoid cleaning the origin of the redirect URL
     * because we want to refresh the page and redirect to the same origin on embedded mode
     */
    yield afterAuthNavigation({
      preventRedirectUrlOriginCleaning: true
    });
    yield put(actions.resetImpersonateState());
    callback == null ? void 0 : callback(true);
  } catch (e) {
    yield put(actions.setImpersonateState({
      loading: false,
      error: errorHandler(e)
    }));
  }
}
export function* impersonateSagas() {
  yield takeLeading(actions.impersonate, impersonate);
}
