// @frontegg/js/node/AdminPortal/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdminPortal = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _AppHolder = require("../AppHolder");
var AdminPortal = /*#__PURE__*/function () {
  function AdminPortal() {
    (0, _classCallCheck2["default"])(this, AdminPortal);
  }
  (0, _createClass2["default"])(AdminPortal, null, [{
    key: "show",
    value: function show() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      _AppHolder.AppHolder.getInstance(name).showAdminPortal();
    }
  }, {
    key: "showMultiApp",
    value: function showMultiApp() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      _AppHolder.AppHolder.getInstance(name).showAdminPortal({
        isMultiApp: true
      });
    }

    /**
     * Open the hosted portal must be called synchronously with user interaction
     *
     * @param newTab - open the portal in a new tab = true
     * @param name - the app name = 'default'
     */
  }, {
    key: "openHosted",
    value: function openHosted() {
      var newTab = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
      var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'default';
      _AppHolder.AppHolder.getInstance(name).openHostedAdminPortal(newTab);
    }
  }, {
    key: "hide",
    value: function hide() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      _AppHolder.AppHolder.getInstance(name).hideAdminPortal();
    }
  }]);
  return AdminPortal;
}();
exports.AdminPortal = AdminPortal;
