// @frontegg/js/AdminPortal/index.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import { AppHolder } from '../AppHolder';
export var AdminPortal = /*#__PURE__*/function () {
  function AdminPortal() {
    _classCallCheck(this, AdminPortal);
  }
  _createClass(AdminPortal, null, [{
    key: "show",
    value: function show() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      AppHolder.getInstance(name).showAdminPortal();
    }
  }, {
    key: "showMultiApp",
    value: function showMultiApp() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      AppHolder.getInstance(name).showAdminPortal({
        isMultiApp: true
      });
    }

    /**
     * Open the hosted portal must be called synchronously with user interaction
     *
     * @param newTab - open the portal in a new tab = true
     * @param name - the app name = 'default'
     */
  }, {
    key: "openHosted",
    value: function openHosted() {
      var newTab = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
      var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'default';
      AppHolder.getInstance(name).openHostedAdminPortal(newTab);
    }
  }, {
    key: "hide",
    value: function hide() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'default';
      AppHolder.getInstance(name).hideAdminPortal();
    }
  }]);
  return AdminPortal;
}();
